### ISSUE TITLE
"Failed to invoke DSC Set method: Authentication needed, call Connect-MgGraph"

### ISSUE THREAD

**Issue by ybsin** [microsoft/Microsoft365DSC#2717]
Hello I am attempting to hook the DSC configurations with Ansible. So my setup is a playbook to copy the ps1 configuration and to send it right at the proper windows server. I then launch the configuration and then add parameters when calling it using the win_dsc module. (This method uses the Invoke DSC module, which a treat and a pain at the same time. )

Now it runs fine if the playbook contains the values as vars. If they come out from a var_files, some other problem will happend in the process. Here's a verbose error stdout of Ansible. 

```bash
 "module_version": "1.22.1221.1",
    "msg": "Failed to invoke DSC Set method: Authentication needed, call Connect-MgGraph.",
    "reboot_required": false,
    "verbose_get": [
        "Perform operation 'Invoke CimMethod' with following parameters, ''methodName' = ResourceGet,'className' = MSFT_DSCLocalConfigurationManager,'namespaceName' = root/Microsoft/Windows/DesiredStateConfiguration'.",
        "An LCM method call arrived from computer WinServ2 with user sid <removed for security>.",
        "[WinServ2]:                            [[AADUser]DirectResourceAccess] Getting configuration of Office 365 User <UserPrincipalName>",
        "[WinServ2]:                            [[AADUser]DirectResourceAccess] Getting Office 365 User <UserPrincipalName>",
        "[WinServ2]:                            [[AADUser]DirectResourceAccess] The specified User doesn't already exist.",
        "[WinServ2]: LCM:  [ End    Get      ]  [[AADUser]DirectResourceAccess]  in 5.4450 seconds.",
        "[WinServ2]: LCM:  [ End    Set      ]    in  5.5540 seconds.",
        "Operation 'Invoke CimMethod' complete.",
        "Time taken for configuration job to complete is 5.815 seconds"
```

I'm on the latest version of the module. As for the Connect-MgGraph comment, it is odd because I did it several times, with several scopes and I don't really see any change. And if I change a few values, this problem disappears. 
Now I understand you're not suited to help me with Ansible, I'm pretty sure the actual problem could be from there, but if you understood the reason behind the problem maybe that'd give me a lead to fix it on there. 
I cannot share the whole thing publicly, but I could send a private playbook for anyone to test with it if necessary, you just need a small windows VM with Vagrant or Azure.

**Comment by andikrueger** [comment]
Could you outline what kind of changes work in favor of the solution?

**Comment by ybsin** [comment]
Apparently, after some more testing. It seems that using credentials from a var_file is the problem. That or the sole fact that I added ansible-vault as encryption to these values. 
 
There is a weird behavior made by the win_dsc plugin as it saves every parameter, authentication methods aside, to generate the DSC, as you have several authentication method in every submodule. Futhermore it logs every credentials or password object as "*_password" so it even changes the name of the DSC parameter in Ansible. This might have something to do in the mix.

My temporary solution is to cut some parameter off the playbook and set the credentials in vars as opposed to var_files but that doesn't make it very dynamic, which would be the end goal..

Either way, I realize my problem isn't up with M365DSC, if you want some more information out of curiosity or for making the same setup I did, I can keep on writing my progress on that matter here.
